src/docs/eos-updater-prepare-volume.8
.\" Manpage for eos-updater-prepare-volume.
.\" Documentation is under the same licence as the eos-updater package.
.TH man 8 "27 Feb 2017" "1.0" "eos\-updater\-prepare\-volume man page"
.\"
.SH NAME
.IX Header "NAME"
eos\-updater\-prepare\-volume — Endless OS Updater USB Drive Preparation Tool
.\"
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.\"
\fBeos\-updater\-prepare\-volume [\-q] \fPvolume-directory\fB
.\"
.SH DESCRIPTION
.IX Header "DESCRIPTION"
.\"
\fBeos\-updater\-prepare\-volume\fP is a utility to prepare a USB drive with a
copy of the computer’s Endless OS updates, so those updates can be applied to
another computer running Endless OS, bringing it up to date.
.PP
The updates will be put in an OSTree repository in the \fBeos\-update\fP
directory on the USB drive. The path of the mounted drive must be passed as the
only argument to the program. Other files on the drive will not be affected.
.PP
\fBeos\-updater\-prepare\-volume\fP is designed to be run manually by system
administrators as a tool for updating computers which otherwise cannot be
updated. It is not intended to be used for routine updates of computers.
.PP
Currently, \fBeos\-updater\-prepare\-volume\fP cannot update an existing
repository on a USB drive — the \fBeos\-update\fP directory must not already
exist on the USB drive.
.PP
If the computer has been converted to not use OSTree, its OS updates cannot be
copied onto a USB drive using this tool.
.\"
.SH OPTIONS
.IX Header "OPTIONS"
.\"
.IP "\fB\-q\fP, \fB\-\-quiet\fP"
Do not print informational output.
.\"
.SH EXAMPLES
.IX Header "EXAMPLES"
.\"
Example of updating \fIold\-machine\fP from \fInew\-machine\fP:
.PP
.nf
.RS
new\-machine$ eos\-updater\-prepare\-volume /run/media/user/some\-volume
.RE
.fi
.PP
On \fIold\-machine\fP, the updates can be applied automatically by running the
following after inserting the USB drive:
.PP
.nf
.RS
old\-machine$ eos\-autoupdater \-\-from\-volume /run/media/user/some\-volume
.RE
.fi
.PP
Alternatively, to do things manually, copy \fIeos\-updater.conf\fP from
\fI/usr/share/eos\-updater/\fP to \fI/etc/eos\-updater/\fP on
\fIold\-machine\fP, and edit it to look like:
.PP
.nf
.RS
[Download]
Order=volume;main;  # add the ‘volume’ source

[Source "volume"]
Path=/run/media/user/some\-volume  # wherever the volume will be mounted
.RE
.fi
.PP
Then run the following commands:
.PP
.nf
.RS
old\-machine$ eos\-updater\-ctl poll
old\-machine$ eos\-updater\-ctl fetch
old\-machine$ eos\-updater\-ctl apply
.RE
.fi
.PP
Finally, reboot \fIold\-machine\fP to finish applying the updates. Remove the
USB drive before or after rebooting.
.\"
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.\"
\fPeos\-updater\-prepare\-volume\fP supports the standard GLib environment
variables for debugging. These variables are \fBnot\fP intended to be used in
production:
.\"
.IP \fI$G_MESSAGES_DEBUG\fP 4
.IX Item "$G_MESSAGES_DEBUG"
This variable can contain one or more debug domain names to display debug output
for. The value \fIall\fP will enable all debug output. The default is for no
debug output to be enabled.
.\"
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
.\"
\fBeos\-updater\-prepare\-volume\fP may return one of several error codes if it
encounters problems.
.\"
.IP "0" 4
.IX Item "0"
No problems occurred. The repository was successfully created on the USB drive.
.\"
.IP "1" 4
.IX Item "1"
There was an error while creating the repository.
.\"
.IP "2" 4
.IX Item "2"
An invalid option was passed to \fBeos\-updater\-prepare\-volume\fP on startup.
.\"
.IP "3" 4
.IX Item "3"
\fBeos\-updater\-prepare\-volume\fP was not run as root.
.\"
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.\"
\fBeos\-autoupdater\fP(8),
\fBeos\-updater\fP(8),
\fBeos\-updater.conf\fP(5),
.\"
.SH BUGS
.IX Header "BUGS"
.\"
Any bugs which are found should be reported on the project website:
.br
\fIhttps://support.endlessm.com/\fP
.\"
.SH AUTHOR
.IX Header "AUTHOR"
.\"
Endless Mobile, Inc.
.\"
.SH COPYRIGHT
.IX Header "COPYRIGHT"
.\"
Copyright © 2017 Endless Mobile, Inc.


debian/changelog
eos-updater (1.3.0~rc3) unstable; urgency=low

  * Updating version to 1.3.0~rc3

 -- Roddy Shuler <roddy@endlessm.com>  Tue, 12 Nov 2013 21:13:35 -0800

eos-updater (1.3.0~rc1) unstable; urgency=low

  * Updating version to 1.3.0~rc1

 -- Roddy Shuler <roddy@endlessm.com>  Fri, 08 Nov 2013 23:38:59 -0800

eos-updater (1.2.1~rc1) unstable; urgency=low

  * Updating version to 1.2.1~rc1

 -- Roddy Shuler <roddy@endlessm.com>  Sun, 27 Oct 2013 22:30:28 -0700

eos-updater (0-1) UNRELEASED; urgency=low

  * Initial release (Closes: #1144)

 -- John Hiesey <john@endlessm.com>  Fri, 04 Oct 2013 11:44:06 -0700


src/eos-updater-prepare-volume.c
/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
 *
 * Copyright 2016 Kinvolk GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * Author: Krzesimir Nowak <krzesimir@kinvolk.io>
 */

#include <glib.h>
#include <locale.h>
#include <ostree.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>

#include "eos-prepare-usb-update.h"

/* main() exit codes. */
enum
{
  EXIT_OK = EXIT_SUCCESS,
  EXIT_FAILED = 1,
  EXIT_INVALID_ARGUMENTS = 2,
  EXIT_RUN_AS_ROOT = 3,
};

static int
fail (gboolean     quiet,
      gint         exit_status,
      const gchar *error_message,
      ...) G_GNUC_PRINTF (3, 4);

static int
fail (gboolean     quiet,
      gint         exit_status,
      const gchar *error_message,
      ...)
{
  va_list ap;
  g_autofree gchar *formatted_message = NULL;

  g_return_val_if_fail (exit_status > 0, EXIT_FAILED);

  if (quiet)
    return exit_status;

  /* Format the arguments. */
  va_start (ap, error_message);
  formatted_message = g_strdup_vprintf (error_message, ap);
  va_end (ap);

  /* Include the usage. */
  g_printerr ("%s: %s\n", g_get_prgname (), formatted_message);

  return exit_status;
}

static int
usage (GOptionContext *context,
       gboolean        quiet,
       const gchar    *error_message,
       ...) G_GNUC_PRINTF (3, 4);

static int
usage (GOptionContext *context,
       gboolean        quiet,
       const gchar    *error_message,
       ...)
{
  va_list ap;
  g_autofree gchar *formatted_message = NULL;
  g_autofree gchar *help = NULL;

  if (quiet)
    return EXIT_INVALID_ARGUMENTS;

  /* Format the arguments. */
  va_start (ap, error_message);
  formatted_message = g_strdup_vprintf (error_message, ap);
  va_end (ap);

  /* Include the usage. */
  help = g_option_context_get_help (context, TRUE, NULL);
  g_printerr ("%s: %s\n\n%s\n", g_get_prgname (), formatted_message, help);

  return EXIT_INVALID_ARGUMENTS;
}

int
main (int argc,
      char **argv)
{
  gboolean quiet = FALSE;
  g_auto(GStrv) remaining = NULL;
  g_autoptr(GOptionContext) context = g_option_context_new ("— Endless OS Updater USB Drive Preparation Tool");
  GOptionEntry entries[] =
    {
      { "quiet", 'q', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &quiet, "Do not print anything; check exit status for success", NULL },
      { G_OPTION_REMAINING, 0, G_OPTION_FLAG_NONE, G_OPTION_ARG_FILENAME_ARRAY,
        &remaining, "Path to the USB drive to prepare", "VOLUME-PATH" },
      { NULL }
    };
  g_autoptr(GError) error = NULL;
  const gchar *raw_usb_path = NULL;
  g_autoptr(GFile) usb_path = NULL;
  g_autoptr(OstreeSysroot) sysroot = NULL;
  g_autoptr(OstreeAsyncProgress) progress = NULL;
  gboolean lock_acquired = FALSE;

  setlocale (LC_ALL, "");

  g_option_context_add_main_entries (context, entries, NULL);
  g_option_context_set_summary (context,
                                "Prepare a USB drive with a copy of the local "
                                "OSTree repository, so it can be used to "
                                "update other machines offline. The repository "
                                "copy will be put in the eos-update directory "
                                "on the USB drive; other files will not be "
                                "affected.");

  if (!g_option_context_parse (context,
                               &argc,
                               &argv,
                               &error))
    {
      return usage (context, quiet, "Failed to parse options: %s",
                    error->message);
    }

  /* We need to be root in order to read all the files in the OSTree repo
   * (unless we’re running the unit tests). */
  if (geteuid () != 0 &&
      g_getenv ("EOS_UPDATER_TEST_UPDATER_DEPLOYMENT_FALLBACK") == NULL)
    {
      return fail (quiet, EXIT_RUN_AS_ROOT, "Must be run as root");
    }

  if (argc != 1 || remaining == NULL || g_strv_length (remaining) != 1)
    {
      return usage (context, quiet,
                    "Expected exactly one path to the USB drive");
    }

  raw_usb_path = remaining[0];
  usb_path = g_file_new_for_commandline_arg (raw_usb_path);
  if (!g_file_query_exists (usb_path, NULL))
    {
      return fail (quiet, EXIT_INVALID_ARGUMENTS, "Path ‘%s’ does not exist",
                   raw_usb_path);
    }

  sysroot = ostree_sysroot_new_default ();

  /* Lock the sysroot so it can’t be updated while we’re pulling from it. The
   * lock is automatically released when we finalise the sysroot. */
  ostree_sysroot_try_lock (sysroot, &lock_acquired, NULL);

  if (!lock_acquired)
    {
      if (!quiet)
        g_print ("Waiting for lock on sysroot…\n");

      if (!ostree_sysroot_lock (sysroot, &error))
        return fail (quiet, EXIT_FAILED, "Failed to lock sysroot: %s",
                     error->message);
    }

  if (!ostree_sysroot_load (sysroot, NULL, &error))
    {
      return fail (quiet, EXIT_FAILED, "Failed to load sysroot: %s",
                   error->message);
    }

  if (!quiet)
    progress = ostree_async_progress_new_and_connect (ostree_repo_pull_default_console_progress_changed,
                                                      /* just whatever that is not NULL, the function
                                                       * above early-quits if user-data is NULL
                                                       */
                                                      main);

  if (!eos_updater_prepare_volume_from_sysroot (sysroot,
                                                usb_path,
                                                progress,
                                                NULL,
                                                &error))
    {
      return fail (quiet, EXIT_FAILED, "Failed to prepare the update: %s",
                   error->message);
    }

  return EXIT_OK;
}
